### PR TITLE
DBZ-1989 Avoid returning distorted (Milli)SecondsBehindSource values

### PR DESCRIPTION
**Pull request by ebrard**
https://issues.jboss.org/browse/DBZ-1989

### CODE REVIEW

**Comment by jpechane** [comment]
Thanks for the PR! I believe there was a misunderstanding. I don't think we should emulate the functuonality for binarly log client - this would not work on JDK 9+.
What we've asked is to use reflection to read the internal data from the stats class and log them so we are able to reason on root cause. The value returned in this case should be the one cached.
This would mean the issue will be bypassed for the future an in the same time it will allow us to gather the necessary insight.

**Comment by ebrard** [comment]
This is what I had in mind at first, but then I thought that between the moment I check internal state and the moment I call `this.stats.getSecondsBehindMaster()`, the internal state might have changed and I might still return a disturbed value instead of the cached one, which is why I thought of just mimicking the function inside Debezium for the moment. I agree that I forgot about the logging part :)

**Comment by gunnarmorling** [comment]
Hey, so I'm a bit confused by all the reflection magic :) What you suggested in the Jira issue made sense to me:

```
    @Override
    public long getMilliSecondsBehindSource() {
        if (this.stats.getSecondsBehindMaster() < SOMELARGEVALUE) {
            return this.stats.getSecondsBehindMaster() * 1000;
        } else {
            return 0;
        }
    }
```

I think returning 0 in this case is as good as returning the last value we returned (it's not the right one either way). A more holistic fix on the Debezium sid would be to calculate the lag always ourselves. For that, `BinlogReader` would have to pass the current event to `BinlogReaderMetrics` (via a new method `onEvent()`) and then we could do the same calcuation as done by `BinaryLogClientStatistics`.

**Comment by ebrard** [comment]
> A more holistic fix on the Debezium sid would be to calculate the lag always ourselves.
Isn't it what this PR does? Well by using reflection, I'll give you that...

**Comment by gunnarmorling** [comment]
Sure, but I'd really prefer hacking into a dependency like that.

**Comment by ebrard** [comment]
Well tbh, for me the comparison with `SOMELARGEVALUE` was just as good, I could make a change in this PR to use this approach instead? I am convinced that it's because of a sync issue in the underlying library (I've opened an issue in there too), so I don't think Debezium should try to address that, checking if the value makes sense and returning zero if it doesn't is fine for me. I've seen in the interface or somewhere that we can also return `-1` in case of weird values, maybe that would be better to return `-1` than `0` since zero has a semantic meaning, whereas `-1` doesn't, what do you think?

**Comment by gunnarmorling** [comment]
> so I don't think Debezium should try to address that

But why not? I mean, the best thing to do would be to fix the underlying issue in the client and update Debezium to that new version. But as it's not quite clear what the fix should be, I think having a meaningful implementation in Debezium which side-steps the issue altogether is the next best thing to do. It should give as an always correct value instead of the occassional 0, -1 or other "fake" value.

**Comment by ebrard** [comment]
Alright, I'll suggest a new code with somehow the approach you suggested (onEvent listener). I am testing it now.

**Comment by ebrard** [comment]
@gunnarmorling I've changed the implementation, can you have a look again?

File: debezium-connector-mysql/src/main/java/io/debezium/connector/mysql/BinlogReader.java [hunk 1]
```diff
@@ -494,6 +495,33 @@ protected void logEvent(Event event) {
...
+        }
+
+        // MySQL has seconds resolution but mysql-binlog-connector-java returns
+        // a value in milliseconds
+        long eventTs = event.getHeader().getTimestamp();
+
+        if (eventTs == 0) {
+            // A zero-valued timestamp can also happen if case of ROTATE events
+            // in which case we keep the previous value
+            logger.trace("Received unexpected event with 0 timestamp: {}", event);
```
Comment (gunnarmorling) [inline debezium-connector-mysql/src/main/java/io/debezium/connector/mysql/BinlogReader.java:516]: It's not so unexpected then as per the comment above?

Comment (ebrard) [reply debezium-connector-mysql/src/main/java/io/debezium/connector/mysql/BinlogReader.java:516]: Well yes but I don't understand why this happens. What I observed are two ROTATE events sequentially received with one of them with no timestamp (they are identical, this aside). And there are plenty of other events so maybe that's not the only case, maybe I should edit a bit the comment?

Comment (gunnarmorling) [reply debezium-connector-mysql/src/main/java/io/debezium/connector/mysql/BinlogReader.java:516]: yes +1; the note to ROTATE is more confusing than anything else then, I think.

**Comment by jpechane** [comment]
@ebrard Hi, I really like the solution? Would it be possible for you to add also an integration test for it?

File: debezium-connector-mysql/src/main/java/io/debezium/connector/mysql/BinlogReader.java (hunk 1 above)
Comment (ebrard) [reply debezium-connector-mysql/src/main/java/io/debezium/connector/mysql/BinlogReader.java:516]: alright, let me modify it a bit

Comment (ebrard) [reply debezium-connector-mysql/src/main/java/io/debezium/connector/mysql/BinlogReader.java:514]: I removed it.

Comment (gunnarmorling) [reply debezium-connector-mysql/src/main/java/io/debezium/connector/mysql/BinlogReader.java:514]: Thx!

**Comment by gunnarmorling** [comment]
Thanks, @ebrard!

**Pull request merged**